File: training-data/python/matrix_chain_multiplication.py
def matrix_chain_order(dims):
    n = len(dims) - 1
    dp = [[0] * n for _ in range(n)]
    split = [[0] * n for _ in range(n)]

    for length in range(2, n + 1):
        for i in range(n - length + 1):
            j = i + length - 1
            dp[i][j] = float("inf")
            for k in range(i, j):
                cost = dp[i][k] + dp[k + 1][j] + dims[i] * dims[k + 1] * dims[j + 1]
                if cost < dp[i][j]:
                    dp[i][j] = cost
                    split[i][j] = k

    return dp[0][n - 1], split


def build_order(split, i, j):
    if i == j:
        return f"M{i}"
    k = split[i][j]
    left = build_order(split, i, k)
    right = build_order(split, k + 1, j)
    return f"({left} x {right})"


if __name__ == "__main__":
    dims = [40, 20, 30, 10, 30]
    min_cost, split = matrix_chain_order(dims)
    print(min_cost)
    print(build_order(split, 0, len(dims) - 2))

File: training-data/php/matrix_chain_multiplication.php
<?php

function matrixChainOrder(array $dims): int
{
    $n = count($dims) - 1;
    $dp = array_fill(0, $n + 1, array_fill(0, $n + 1, 0));

    for ($len = 2; $len <= $n; $len++) {
        for ($i = 1; $i <= $n - $len + 1; $i++) {
            $j = $i + $len - 1;
            $dp[$i][$j] = PHP_INT_MAX;
            for ($k = $i; $k < $j; $k++) {
                $cost = $dp[$i][$k] + $dp[$k + 1][$j] + $dims[$i - 1] * $dims[$k] * $dims[$j];
                if ($cost < $dp[$i][$j]) {
                    $dp[$i][$j] = $cost;
                }
            }
        }
    }
    return $dp[1][$n];
}

// dims describes matrices of size dims[i-1] x dims[i]
$dims = [40, 20, 30, 10, 30];
echo matrixChainOrder($dims) . "\n";
